apps/src/code-studio/initApp/loadApp.js

```diff
@@ -30,7 +40,181 @@ function showDisabledButtonsAlert(isHocScript) {
+        ReactDOM.render(
+          <PlayZone
+            stageName={stageName}
+            onContinue={() => { dialog.hide(); }}
+            i18n={window.dashboard.i18n}
+          />,
+          body
+        );
+        const dialog = new Dialog({
+          body: body,
+          width: 800,
+          redirect: lastServerResponse.nextRedirect
+        });
+        dialog.show();
+      } else if (lastServerResponse.nextRedirect) {
+        window.location.href = lastServerResponse.nextRedirect;
+      }
+    },
+    backToPreviousLevel: function () {
+      var lastServerResponse = reporting.getLastServerResponse();
+      if (lastServerResponse.previousLevelRedirect) {
+        window.location.href = lastServerResponse.previousLevelRedirect;
+      }
+    },
+    showInstructionsWrapper: function (showInstructions) {
+      // Always skip all pre-level popups on share levels or when configured thus
+      if (this.share || appOptions.level.skipInstructionsPopup) {
+        return;
+      }
+
+      var afterVideoCallback = showInstructions;
+      if (appOptions.level.afterVideoBeforeInstructionsFn) {
+        afterVideoCallback = function () {
+          appOptions.level.afterVideoBeforeInstructionsFn(showInstructions);
+        };
+      }
+
+      var hasVideo = !!appOptions.autoplayVideo;
+      var hasInstructions = !!(appOptions.level.instructions ||
+                               appOptions.level.aniGifURL);
+      var noAutoplay = !!queryString.parse(location.search).noautoplay;
+
+      if (hasVideo && !noAutoplay) {
+        if (hasInstructions) {
+          appOptions.autoplayVideo.onClose = afterVideoCallback;
+        }
+        showVideoDialog(appOptions.autoplayVideo);
+      } else {
+        if (hasVideo && noAutoplay) {
+          clientState.recordVideoSeen(appOptions.autoplayVideo.key);
+        }
+        if (hasInstructions) {
+          afterVideoCallback();
+        }
+      }
+    }
+  };
+  $.extend(true, appOptions, baseOptions);
+
+  // Turn string values into functions for keys that begin with 'fn_' (JSON can't contain function definitions)
+  // E.g. { fn_example: 'function () { return; }' } becomes { example: function () { return; } }
+  (function fixUpFunctions(node) {
+    if (typeof node !== 'object') {
+      return;
+    }
+    for (var i in node) {
+      if (/^fn_/.test(i)) {
+        try {
+          // eslint-disable-next-line no-eval
+          node[i.replace(/^fn_/, '')] = eval('(' + node[i] + ')');
+        } catch (e) {
+        }
+      } else {
+        fixUpFunctions(node[i]);
+      }
+    }
+  })(appOptions.level);
+
+  // Previously, this was set by dashboard based on route and user agent. We
+  // stopped being able to use the user agent on the server, and thus try
+  // to have the same logic on the client.
+  appOptions.noPadding = userAgentParser.isMobile();
+}
+
+function loadAppAsync(appOptions, callback) {
+  setupApp(appOptions);
```

Comment:
In loadAppAsync? It looks like a bunch of different things, mostly related to progress tracking? Was there something specific you had questions about?